File: src/org/antlr/runtime/debug/DebugEventSocketProxy.java
/*
 [The "BSD licence"]
 Copyright (c) 2005 Terence Parr
 All rights reserved.

 Redistribution and use in source and binary forms, with or without
 modification, are permitted provided that the following conditions
 are met:
 1. Redistributions of source code must retain the above copyright
    notice, this list of conditions and the following disclaimer.
 2. Redistributions in binary form must reproduce the above copyright
    notice, this list of conditions and the following disclaimer in the
    documentation and/or other materials provided with the distribution.
 3. The name of the author may not be used to endorse or promote products
    derived from this software without specific prior written permission.

 THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/
package org.antlr.runtime.debug;

import org.antlr.runtime.*;

import java.net.ServerSocket;
import java.net.Socket;
import java.io.*;

public class DebugEventSocketProxy implements DebugEventListener {
	public static final int DEFAULT_DEBUGGER_PORT = 2005;
	protected int port = DEFAULT_DEBUGGER_PORT;
	protected ServerSocket serverSocket;
	protected Socket socket;
	PrintWriter out;
	BufferedReader in;

	public DebugEventSocketProxy() {
		this(DEFAULT_DEBUGGER_PORT);
	}

	public DebugEventSocketProxy(int port) {
		this.port = port;
	}

	public void handshake() throws IOException {
		if ( serverSocket==null ) {
			serverSocket = new ServerSocket(port);
			socket = serverSocket.accept();
			socket.setTcpNoDelay(true);
			OutputStream os = socket.getOutputStream();
			OutputStreamWriter osw = new OutputStreamWriter(os, "UTF8");
			out = new PrintWriter(new BufferedWriter(osw));
			InputStream is = socket.getInputStream();
			InputStreamReader isr = new InputStreamReader(is, "UTF8");
			in = new BufferedReader(isr);
			transmit("ANTLR 3.0ea1");
		}
	}

	public void commence() {
		// don't bother sending event; listener will trigger upon connection
	}

	public void terminate() {
		transmit("terminate");
		out.close();
		try {
			socket.close();
		}
		catch (IOException ioe) {
			ioe.printStackTrace(System.err);
		}
	}

	protected void ack() {
		try {
			String ack = in.readLine();
		}
		catch (IOException ioe) {
			ioe.printStackTrace(System.err);
		}

	}

	protected void transmit(String event) {
		out.println(event);
		out.flush();
		ack();
	}

	public void enterRule(String ruleName) {
		transmit("enterRule "+ruleName);
	}

	public void enterAlt(int alt) {
		transmit("enterAlt "+alt);
	}

	public void exitRule(String ruleName) {
		transmit("exitRule "+ruleName);
	}

	public void enterSubRule(int decisionNumber) {
		transmit("enterSubRule "+decisionNumber);
	}

	public void exitSubRule(int decisionNumber) {
		transmit("exitSubRule "+decisionNumber);
	}

	public void enterDecision(int decisionNumber) {
		transmit("enterDecision "+decisionNumber);
	}

	public void exitDecision(int decisionNumber) {
		transmit("exitDecision "+decisionNumber);
	}

	public void consumeToken(Token t) {
		String buf = serializeToken(t);
		transmit("consumeToken "+buf);
	}

	public void consumeHiddenToken(Token t) {
		String buf = serializeToken(t);
		transmit("consumeHiddenToken "+buf);
	}

	public void LT(int i, Token t) {
        if(t != null)
            transmit("LT "+i+" "+serializeToken(t));
	}

	public void mark(int i) {
		transmit("mark "+i);
	}

	public void rewind(int i) {
		transmit("rewind "+i);
	}

	public void location(int line, int pos) {
		transmit("location "+line+" "+pos);
	}

	public void recognitionException(RecognitionException e) {
		StringBuffer buf = new StringBuffer(50);
		buf.append("exception ");
		buf.append(e.getClass().getName());
		// dump only the data common to all exceptions for now
		buf.append(" ");
		buf.append(e.index);
		buf.append(" ");
		buf.append(e.line);
		buf.append(" ");
		buf.append(e.charPositionInLine);
		transmit(buf.toString());
	}

	public void beginResync() {
		transmit("beginResync");
	}

	public void endResync() {
		transmit("endResync");
	}

	public void semanticPredicate(boolean result, String predicate) {
		predicate = escapeNewlines(predicate);
		StringBuffer buf = new StringBuffer(50);
		buf.append("semanticPredicate ");
		buf.append(result);
		buf.append(" ");
		buf.append(predicate);
		transmit(buf.toString());
	}

	protected String serializeToken(Token t) {
		StringBuffer buf = new StringBuffer(50);
		buf.append(t.getTokenIndex()); buf.append(' ');
		buf.append(t.getType()); buf.append(' ');
		buf.append(t.getChannel()); buf.append(' ');
		buf.append(t.getLine()); buf.append(' ');
		buf.append(t.getCharPositionInLine()); buf.append(" \"");
		String txt = t.getText();
		if ( txt==null ) {
			txt = "";
		}
		// escape \n and \r all text for token appears to exist on one line
		// this escape is slow but easy to understand
		txt = escapeNewlines(txt);
		buf.append(txt);
		return buf.toString();
	}

	protected String escapeNewlines(String txt) {
		txt = txt.replaceAll("%","%25");   // escape all escape char ;)
		txt = txt.replaceAll("\n","%0A");  // escape \n
		txt = txt.replaceAll("\r","%0D");  // escape \r
		return txt;
	}
}

File: runtime/CSharp/Antlr3.Runtime/Antlr.Runtime.Debug/DebugEventSocketProxy.cs
namespace Antlr.Runtime.Debug
{
    using Antlr.Runtime;
    using System;
    using System.IO;
    using System.Net.Sockets;
    using System.Text;

    public class DebugEventSocketProxy : BlankDebugEventListener
    {
        public const int DEFAULT_DEBUGGER_PORT = 0xc001;
        protected string grammarFileName;
        protected int port;
        protected StreamReader reader;
        protected TcpListener serverSocket;
        protected TcpClient socket;
        protected StreamWriter writer;

        public DebugEventSocketProxy() : this(null, 0xc001)
        {
        }

        public DebugEventSocketProxy(string grammarFileName, int port)
        {
            this.port = 0xc001;
            this.grammarFileName = grammarFileName;
            this.port = port;
        }

        protected internal virtual void Ack()
        {
            try
            {
                this.reader.ReadLine();
            }
            catch (IOException exception)
            {
                Console.Error.WriteLine(exception.StackTrace);
            }
        }

        public override void AddChild(int rootID, int childID)
        {
            this.Transmit(string.Concat(new object[] { "AddChild ", rootID, " ", childID }));
        }

        public override void BecomeRoot(int newRootID, int oldRootID)
        {
            this.Transmit(string.Concat(new object[] { "BecomeRoot ", newRootID, " ", oldRootID }));
        }

        public override void BeginBacktrack(int level)
        {
            this.Transmit("BeginBacktrack " + level);
        }

        public override void BeginResync()
        {
            this.Transmit("BeginResync");
        }

        public override void Commence()
        {
        }

        public override void ConsumeHiddenToken(IToken t)
        {
            this.Transmit("ConsumeHiddenToken " + this.SerializeToken(t));
        }

        public override void ConsumeNode(int ID, string text, int type)
        {
            text = this.EscapeNewlines(text);
            StringBuilder builder = new StringBuilder(50);
            builder.Append("ConsumeNode ");
            builder.Append(ID);
            builder.Append(" ");
            builder.Append(type);
            builder.Append(" ");
            builder.Append(text);
            this.Transmit(builder.ToString());
        }

        public override void ConsumeToken(IToken t)
        {
            this.Transmit("ConsumeToken " + this.SerializeToken(t));
        }

        public override void CreateNode(int ID, int tokenIndex)
        {
            this.Transmit(string.Concat(new object[] { "CreateNode ", ID, " ", tokenIndex }));
        }

        public override void CreateNode(int ID, string text, int type)
        {
            text = this.EscapeNewlines(text);
            StringBuilder builder = new StringBuilder(50);
            builder.Append("CreateNodeFromToken ");
            builder.Append(ID);
            builder.Append(" ");
            builder.Append(type);
            builder.Append(" ");
            builder.Append(text);
            this.Transmit(builder.ToString());
        }

        public override void EndBacktrack(int level, bool successful)
        {
            bool flag;
            this.Transmit(string.Concat(new object[] { "EndBacktrack ", level, " ", successful ? (flag = true).ToString() : (flag = false).ToString() }));
        }

        public override void EndResync()
        {
            this.Transmit("EndResync");
        }

        public override void EnterAlt(int alt)
        {
            this.Transmit("EnterAlt " + alt);
        }

        public override void EnterDecision(int decisionNumber)
        {
            this.Transmit("EnterDecision " + decisionNumber);
        }

        public override void EnterRule(string ruleName)
        {
            this.Transmit("EnterRule " + ruleName);
        }

        public override void EnterSubRule(int decisionNumber)
        {
            this.Transmit("EnterSubRule " + decisionNumber);
        }

        protected internal virtual string EscapeNewlines(string txt)
        {
            txt = txt.Replace("%", "%25");
            txt = txt.Replace("\n", "%0A");
            txt = txt.Replace("\r", "%0D");
            return txt;
        }

        public override void ExitDecision(int decisionNumber)
        {
            this.Transmit("ExitDecision " + decisionNumber);
        }

        public override void ExitRule(string ruleName)
        {
            this.Transmit("ExitRule " + ruleName);
        }

        public override void ExitSubRule(int decisionNumber)
        {
            this.Transmit("ExitSubRule " + decisionNumber);
        }

        public override void GetNilNode(int ID)
        {
            this.Transmit("GetNilNode " + ID);
        }

        public virtual void Handshake()
        {
            if (this.serverSocket == null)
            {
                this.serverSocket = new TcpListener(this.port);
                this.serverSocket.Start();
                this.socket = this.serverSocket.AcceptTcpClient();
                this.socket.NoDelay = true;
                this.reader = new StreamReader(this.socket.GetStream(), Encoding.UTF8);
                this.writer = new StreamWriter(this.socket.GetStream(), Encoding.UTF8);
                this.writer.WriteLine("ANTLR " + Constants.DEBUG_PROTOCOL_VERSION);
                this.writer.WriteLine("grammar \"" + this.grammarFileName);
                this.writer.Flush();
            }
        }

        public override void Location(int line, int pos)
        {
            this.Transmit(string.Concat(new object[] { "Location ", line, " ", pos }));
        }

        public override void LT(int i, IToken t)
        {
            if (t != null)
            {
                this.Transmit(string.Concat(new object[] { "LT ", i, " ", this.SerializeToken(t) }));
            }
        }

        public override void LT(int i, int ID, string text, int type)
        {
            text = this.EscapeNewlines(text);
            StringBuilder builder = new StringBuilder(50);
            builder.Append("LN ");
            builder.Append(i);
            builder.Append(" ");
            builder.Append(ID);
            builder.Append(" ");
            builder.Append(type);
            builder.Append(" ");
            builder.Append(text);
            this.Transmit(builder.ToString());
        }

        public override void Mark(int i)
        {
            this.Transmit("Mark " + i);
        }

        public override void RecognitionException(Antlr.Runtime.RecognitionException e)
        {
            StringBuilder builder = new StringBuilder(50);
            builder.Append("Exception ");
            builder.Append(e.GetType().FullName);
            builder.Append(" ");
            builder.Append(e.Index);
            builder.Append(" ");
            builder.Append(e.Line);
            builder.Append(" ");
            builder.Append(e.CharPositionInLine);
            this.Transmit(builder.ToString());
        }

        public override void Rewind()
        {
            this.Transmit("Rewind");
        }

        public override void Rewind(int i)
        {
            this.Transmit("Rewind " + i);
        }

        public override void SemanticPredicate(bool result, string predicate)
        {
            predicate = this.EscapeNewlines(predicate);
            StringBuilder builder = new StringBuilder(50);
            builder.Append("SemanticPredicate ");
            builder.Append(result);
            builder.Append(" ");
            builder.Append(predicate);
            this.Transmit(builder.ToString());
        }

        protected internal virtual string SerializeToken(IToken t)
        {
            StringBuilder builder = new StringBuilder(50);
            builder.Append(t.TokenIndex);
            builder.Append(' ');
            builder.Append(t.Type);
            builder.Append(' ');
            builder.Append(t.Channel);
            builder.Append(' ');
            builder.Append(t.Line);
            builder.Append(' ');
            builder.Append(t.CharPositionInLine);
            builder.Append(" \"");
            string text = t.Text;
            if (text == null)
            {
                text = "";
            }
            text = this.EscapeNewlines(text);
            builder.Append(text);
            return builder.ToString();
        }

        public override void SetTokenBoundaries(int ID, int tokenStartIndex, int tokenStopIndex)
        {
            this.Transmit(string.Concat(new object[] { "SetTokenBoundaries ", ID, " ", tokenStartIndex, " ", tokenStopIndex }));
        }

        public override void Terminate()
        {
            this.Transmit("Terminate");
            this.writer.Close();
            try
            {
                this.socket.Close();
            }
            catch (IOException exception)
            {
                Console.Error.WriteLine(exception.StackTrace);
            }
        }

        protected internal virtual void Transmit(string eventLabel)
        {
            this.writer.WriteLine(eventLabel);
            this.writer.Flush();
            this.Ack();
        }
    }
}

